src/coreclr/jit/lowerxarch.cpp

```diff
@@ -1314,10 +1314,14 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
                 // the zmask from op1. We expect that op2 has already been
                 // lowered and therefore the containment checks have happened
 
+                // Since this is a newer operation, we need to account for
+                // the possibility of `op1Intrinsic` zeroing the same element
+                // we're setting here.
+
                 assert(op1Intrinsic->Op(2)->isContained());
 
                 ssize_t op1Ival = op1Idx->AsIntConCommon()->IconValue();
-                ival |= (op1Ival & 0x0F);
+                ival |= ((op1Ival & 0x0F) & ((1 << count_d) - 1));
```

Comment:
We have `Insert(Insert(vector, data1, op1Ival), data2, ival)`

That is, `ival` is the immediate of the outer `Insert` and `op1Ival` is the immediate of the inner `Insert`.

The `ival` is broken up into:
```
zmask: 4-bits
count_d: 2-bits
count_s: 2-bits
```

The `zmask` indicates which elements in the result should be set to zero, this overrides any other bits.
The `count_d` indicates the index of the result that is written using the value at the index of the input (`count_s`)

So this was taking the `zmask` from `op1IVal` and ensuring that the same bits are being zeroed in the new `ival`.

-------------------------

Given a scenario that looks like
```csharp
Vector128<float> Example(Vector128<float> input)
{
    return input.WithElement(2, 0.0f)
                .WithElement(3, 1.0f);
}
```

This ends up lowering down to:
```csharp
Vector128<float> tmp = Sse41.Insert(input, Vector128<float>.Zero, 0x20);
return Sse41.Insert(tmp, Vector128.Create(1.0f), 0x30);
```

Since insert of `Zero` is special, we end up changing this to:
```csharp
Vector128<float> tmp = Sse41.Insert(input, Vector128<float>.Zero, 0x24); // Zero is contained
return Sse41.Insert(tmp, Vector128.Create(1.0f), 0x30);
```

We then can further optimize this by folding the two operations into one by combining the two zmasks:
```csharp
return Sse41.Insert(input, Vector128.Create(1.0f), 0x34);
```

This allows us to fold any number of `WithElement(n, 0.0f)` in a chain of `Inserts` to reduce the number of operations.

------------------------

The bug here is one where we specifically had:
```csharp
Vector128<float> Example(Vector128<float> input)
{
    return input.WithElement(3, 0.0f)
                .WithElement(3, 1.0f);
}
```

Which lowered to:
```csharp
Vector128<float> tmp = Sse41.Insert(input, Vector128<float>.Zero, 0x30);
return Sse41.Insert(tmp, Vector128.Create(1.0f), 0x30);
```

And then became:
```csharp
Which lowered to:
```csharp
Vector128<float> tmp = Sse41.Insert(input, Vector128<float>.Zero, 0x38);
return Sse41.Insert(tmp, Vector128.Create(1.0f), 0x30);
```

Since we are going from `inner: zero`, `outer: non-zero`, we need to handle the case where `outer` is setting a value that a previous `Insert` was zeroing and so we need to ensure we take the `zmask` from `op1IVal`, but mask off the element we're setting as part of the new `ival`.

Given we're `or'ing` the bits together, we'll still preserve any zeroing that `ival` is doing itself.

Reply:
That being said, I just realized this should be:
```diff
- ival |= ((op1Ival & 0x0F) & ((1 << count_d) - 1));
+ ival |= ((op1Ival & 0x0F) & ~(1 << count_d));
```

Otherwise we'd mess up for:
```csharp
Vector128<float> Example(Vector128<float> input)
{
    return input.WithElement(2, 0.0f)
                .WithElement(3, 0.0f)
                .WithElement(2, 1.0f);
}
```

Since that'd be `0x24`, `0x38`, and `0x20`, which would become `0x3C` and `0x20`, and then `0x24` when it should become `0x28`